mmv1/products/cloudquotas/QuotaPreference.yaml

```diff
@@ -0,0 +1,164 @@
+# limitations under the License.
+
+--- !ruby/object:Api::Resource
+name: 'QuotaPreference'
+description: |
+  QuotaPreference represents the preferred quota configuration specified for a project, folder or organization. There is only one QuotaPreference resource for a quota value targeting a unique set of dimensions.
+references: !ruby/object:Api::Resource::ReferenceLinks
+  guides:
+    'Cloud Quotas Overview': 'https://cloud.google.com/docs/quotas/overview'
+  api: 'https://cloud.google.com/docs/quotas/reference/rest/v1/projects.locations.quotaPreferences'
+base_url: '{{parent}}/locations/global/quotaPreferences'
+self_link: '{{parent}}/locations/global/quotaPreferences/{{quota_preference_id}}'
+create_url: '{{parent}}/locations/global/quotaPreferences?quotaPreferenceId={{quota_preference_id}}&ignoreSafetyChecks={{ignore_safety_checks}}'
+update_url: '{{parent}}/locations/global/quotaPreferences/{{quota_preference_id}}?allowMissing={{allow_missing}}&validateOnly={{validate_only}}&ignoreSafetyChecks={{ignore_safety_checks}}'
+update_verb: :PATCH
+update_mask: true
+skip_delete: true
+autogen_async: false
+custom_code: !ruby/object:Provider::Terraform::CustomCode
+  encoder: templates/terraform/encoders/quota_preference.go.erb
+
+examples:
+  - !ruby/object:Provider::Terraform::Examples
+    name: "cloudquotas_quota_preference_basic"
+    primary_resource_id: "preference"
+    skip_test: true
+    vars:
+      quota_preference_id: "compute_googleapis_com-CPUS-per-project_us-central2"
+    test_env_vars:
+      project: :PROJECT_NAME
+
+parameters:
+  - !ruby/object:Api::Type::String
+    name: 'parent'
+    immutable: true
+    url_param_only: true
+    description: The parent of the quota preference. Allowed parents are "projects/[project-id / number]" or "folders/[folder-id / number]" or "organizations/[org-id / number]".
+  - !ruby/object:Api::Type::String
+    name: 'quota_preference_id'
+    immutable: true
+    url_param_only: true
+    default_value: ""
+    description: The unique quota preference id.
+  - !ruby/object:Api::Type::Boolean
+    name: 'allow_missing'
+    url_param_only: true
+    description: |
+      If set to true, and the quota preference is not found, a new one will be created. In this situation, updateMask is ignored.
+  - !ruby/object:Api::Type::Boolean
+    name: 'validate_only'
+    url_param_only: true
+    description: |
+      If set to true, validate the request, but do not actually update. Note that a request being valid does not mean that the request is guaranteed to be fulfilled.
+  - !ruby/object:Api::Type::Enum
+    name: 'ignore_safety_checks'
+    url_param_only: true
+    default_value: :QUOTA_SAFETY_CHECK_UNSPECIFIED
+    description: The list of quota safety checks to be ignored.
+    values:
+      - :QUOTA_SAFETY_CHECK_UNSPECIFIED
+      - :QUOTA_DECREASE_BELOW_USAGE
+      - :QUOTA_DECREASE_PERCENTAGE_TOO_HIGH
+
+properties:
+  - !ruby/object:Api::Type::String
+    name: 'name'
```

Comment:
If I'm understanding correctly, using `name` takes precedence over `parent` and `quota_preference_id`, even if they are specified. Is that correct? If so, we may want to consider using `conflicts_with` or similar to prevent unexpected behavior.

Reply:
Yes I was thinking `parent` and `quota_preference_id` fields will be fetched from `name` if none are specified. Also name is a required field for UPDATE, so the user do not need to input redundant information. I've added `conflicts_with` for `name`.

Reply:
Ok, the "conflicts with" looks good. It may still be worth reducing the duplication, and I'm remembering that we usually use a pattern here where `name` doubles as the name and the id. See how Cloud Run V2 Service is set up with https://cloud.google.com/run/docs/reference/rest/v2/projects.locations.services/create and https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/cloudrunv2/Service.yaml#L135-L145. The user would specify the `parent` and optionally the `name`, and after creation, the `name` would contain the full name value including the parent.

```
resource "google_cloud_quotas_quota_preference" "my-preference" {
  parent = "projects/123/locations/global"
  name = "compute_googleapis_com-CPUS-per-project-us-central1"
  ...
```

After apply, the `name` in TF state is `"projects/123/locations/global/quotaPreferences/compute_googleapis_com-CPUS-per-project-us-central1"`.

Reply:
Implementation updated to match the example's approach.